Netcracker/KubeMarine
# [CPDEV-87623] Unify enrichment, installation and upgrade of sandbox_image for public and private environments

### Description
* Kubemarine installs different version of `sandbox_image` for containerd for public and private environments (i. e. where `registry` section is correspondingly absent and present).
* Kubemarine does not take into account custom `sandbox_image` during upgrade.

Fixes #465

### Solution
* Added default value of `kubeadm.imageRepository` to `registry.k8s.io`. Values of `sandbox_image`, `kubeadm.dns.imageRepository` are enriched and installed even on public clusters. This ensures the same version of `sandbox_image` by default in both modes.
* Enrichment of `sandbox_image` during upgrade now follows the same rule as enrichment of plugins and packages. Added section in *procedure inventory* to control upgrade of `sandbox_image`. If a user specified custom `sandbox_image` in the inventory, they should provide new version in the procedure inventory that will be taken into account.
* Configuration of containerd during upgrade now happens within `kubernetes` task. `upgrade_containerd` task is removed.
* Added patch that installs new version of the image to match new enrichment of inventory (applicable primarily to public clusters).
* Added task in `check_paas` to check configuration of containerd.
* Minor fixes in the registry enrichment.

### How to apply
Run `kubemarine migrate_kubemarine --force-apply manage_containerd_sandbox_image`.

### Test Cases

**TestCase 1**

Install public cluster.

1. Install public cluster (without `registry` section) of version v1.24.2 on containerd CRI.
2. Check factually used pause version.
   ```
   ps aux | grep "/pause$" | awk '{print $2}'
   sudo /proc/<PID>/exe -v
   ```

| Before | After |
| ------ | ------ |
| [Default 3.6](https://github.com/containerd/containerd/blob/v1.6.12/docs/cri/config.md?plain=1#L43) image is used. | [Enriched 3.7](https://github.com/Netcracker/KubeMarine/blob/0.20.0/kubemarine/resources/configurations/compatibility/internal/kubernetes_images.yaml#L155) image is used |

**TestCase 2**

Validation of `sandbox_image` during upgrade.

1. Specify custom `sandbox_image: example.com:<port>/pause:3.7`
2. Install cluster of version v1.24.2 on containerd CRI no matter public or private.

AR=ER: `/etc/containerd/config.toml` has custom image.

3. Upgrade to v1.25.2.

| Before | After |
| ------ | ------ |
| `/etc/containerd/config.toml` has `example.com:<port>/pause:3.8` image, but `cluster_finalized.yaml`  and new `cluster.yaml` still contain old 3.7 version. | Upgrade fails with suggestion to specify new custom image |

**TestCase 3**

Control upgrade of `sandbox_image`.

1. Specify custom `sandbox_image: example.com:<port>/pause:3.7`
2. Install cluster of version v1.24.2 on containerd CRI no matter public or private.
3. Upgrade to v1.25.2 with new `sandbox_image: example.com:<port>/pause:3.8` in the procedure inventory.

ER: `/etc/containerd/config.toml`, `/var/lib/kubelet/kubeadm-flags.env`, `cluster_finalized.yaml`, and new `cluster.yaml` have new 3.8 version.

**TestCase 4**

Migration of old clusters.

1. Install public cluster (without `registry` section) of version v1.24.2 on containerd CRI **using older version of Kubemarine**.
2. Run check_paas using new version of Kubemarine.

ER: `services.container_runtime.configuration` task fails.

3. Run `migrate_kubemarine --force-apply manage_containerd_sandbox_image` using new version of Kubemarine.

ER: `/etc/containerd/config.toml`, `/var/lib/kubelet/kubeadm-flags.env` have image `registry.k8s.io/pause:3.7`

4. Run check_paas using new version of Kubemarine.

ER: `services.container_runtime.configuration` task succeeds.

### Checklist

#### Unit tests
test_registry.py - added tests to cover registry enrichment.
test_upgrade.py - added tests to cover upgrade enrichment of `sandbox_image`.
